File: 572 Subtree of Another Tree/Subtree of Another Tree.cpp
class Solution {
public:
    bool isSubtree(TreeNode* full, TreeNode* sub) {
        if (sub == nullptr) {
            return true;
        }
        if (full == nullptr) {
            return false;
        }
        if (isSame(full, sub)) {
            return true;
        }

        return isSubtree(full->left, sub) || isSubtree(full->right, sub);
    }

    bool isSame(TreeNode* full, TreeNode* sub) {
        if (full == nullptr && sub == nullptr) {
            return true;
        }
        else if (full != nullptr && sub != nullptr) {
            if (full->val == sub->val) {
                return isSame(full->left, sub->left) && isSame(full->right, sub->right);
            }
        }
        return false;
    }
};

File: 572 Subtree of Another Tree/Subtree of Another Tree.py
#Python3 versiyonunu seçmeyi unutma =)
class Solution:
    def isSubtree(self, full: Optional[TreeNode], sub: Optional[TreeNode]) -> bool:
        if not sub:
            return True
        if not full:
            return False
        if self.isSame(full, sub):
            return True

        return (self.isSubtree(full.left, sub) or self.isSubtree(full.right, sub))

    def isSame(self, full, sub):
        if not full and not sub:
            return True
        elif full and sub:
            if full.val == sub.val:
                return (self.isSame(full.left, sub.left) and self.isSame(full.right, sub.right))
        return False
